doc/cascadia/Cascading-Default-Settings.md

```diff
@@ -0,0 +1,399 @@
+
+### Goal: Minimize Writing to `profiles.json`
+
+We want to write the user settings file, `profiles.json`, as little as possible.
+Each time we write the file, there's the possiblity that we've re-ordered the
+keys, as `jsoncpp` provides no ordering guarantee of the keys. This isn't great,
+as each write of the file will randomly re-order the file.
+
+One of our overarching goals with this change should be to re-write the user
+settings file as little as possible.
+
+## Solution Design
+
+The settings are now composed from two files: a "Default" settings file, and a
+"User" settings file.
+
+### Default Settings
+
+We'll have a static version of the "Default" file hardcoded within the
+application. We'll write the hardcoded version of this file out to an actual
+file on disk, `defaults.json`, however, that file will not actually be used. The
+defaults file will contain some sort of disclaimer like:
+
+```json
+// This is an auto-generated file. Place any modifications to your settings in "profiles.json"
+```
+
+This disclaimer will help identify that the file shouldn't be modified.
+
+The `defaults.json` file will be written out whenever we detect that it doesn't exist.
+
+**TODO**: We also need to write it out if it changes, though ideally without
+actually loading the file at first load. Perhaps we could start another
+thread/`fire_and_forget` once we're loaded to re-scan the file and re-write it
+if it has changed.
+
+Because the `defaults.json` file is hardcoded within our application, we can use
+its text directly, without loading the file from disk. This should help save
+some startup time, as we'll only need to load the user settings from disk.
+
+When we make changes to the default settings, or we make changes to the settings
+schema, we should make sure that we update the hardcoded `defaults.json` with
+the new values. That way, the `defaults.json` file will always have the complete
+set of setings in it.
+
+
+### Layering settings
+
+When we load the settings, we'll do it in two stages. First, we'll deseriale the
+default settings that we've hardcoded. Then, we'll intelligently layer the
+user's setting upon those we've already loaded. Some objects, like the default
+profiles, we'll need to make sure to load from the user settings into the
+existing objects we created from the default settings.
+
+* We'll need to make sure that any profile in the user settings that has a GUID
+  matching a default profile loads the user settings into the object created
+  from the defaults.
+* We'll need to make sure that there's only one action bound to each key chord
+  for a keybinding. If there are any key chords in the user settings that match
+  a default key chord, we should bind them to the action from the user settings
+  instead.
+* For any color schemes who's name matches the name of a default color scheme,
+  we'll need to apply the user settings to the existing color scheme.
+

```

Comment:
Design q: Should we have a `source` attribute for every setting? That's a _huge overhead_ but it lets us know what's really going on. Imagine:

`Source: default`
`Source: user profile, profiles.json`
`Source: user profile, schemes.json`
`Source: runtime` (like customizations applied from VT)

Reply:
This could  help us know which to serialize where

Reply:
That seems like an immense amount of overhead. I know that Michael and Paul discussed doing something similar for conhost back in the day, but that was when we had settings scattered across the registry, in shortcuts, and in runtime args. Considering we're proposing 2 settings files (one that'd be a static known entity), I don't think this is as much of a problem.

Reply:
(I'm thinking this should be resolve as won't-fix? I could add a segment to the "Future Considerations" section if you want)

Reply:
#wontfix